src/serprog.c

```diff
@@ -0,0 +1,510 @@
+        "please use the format /dev/ttyACM0\n";
+    char port_default[] = "/dev/ttyACM0";
+    char *serialdev;
+    char *port = mmt_strdup(pt);
+
+    if (str_eq(port, "unknown")) {
+        port = port_default;
+    }
+
+    serialdev = strtok(port, ",");
+    if (!serialdev) {
+        pmsg_error("%s", port_error);
+        return -1;
+    }
+
+    union pinfo pinfo;
+    pgm->port = port;
+    pinfo.serialinfo.baud = pgm->baudrate? pgm->baudrate: 115200;
+    pinfo.serialinfo.cflags = SERIAL_8N1;
+    if (serial_open(port, pinfo, &pgm->fd)==-1) {
+        return -1;
+    }
+
+    unsigned char buf[32];
+
+    // sync
+    memset(buf, 0, sizeof(buf));
+    if (perform_serprog_cmd(pgm, S_CMD_SYNCNOP, NULL, 0, buf, 1) != 1 || buf[0] != S_ACK) {
+        pmsg_error("cannot sync; is this a serprog programmer?\n");
+        return -1;
+    }
+
+    // get command bitmap
+    memset(my.cmd_bitmap, 0, sizeof(my.cmd_bitmap));
+    if (perform_serprog_cmd(pgm, S_CMD_Q_CMDMAP, NULL, 0, my.cmd_bitmap, 32) != 0) {
+        pmsg_error("cannot get list of supported serprog commands\n");
+        return -1;
+    }
+
+    // get protocol version
+    memset(buf, 0, sizeof(buf));
+    if (!is_serprog_cmd_supported(my.cmd_bitmap, S_CMD_Q_IFACE)
+        || perform_serprog_cmd(pgm, S_CMD_Q_IFACE, NULL, 0, buf, 2) != 0) {
+        pmsg_error("cannot get serprog protocol version\n");
+        return -1;
+    }
+    if (read_le16(buf) != 0x01) {
+        pmsg_error("unsupported serprog protocol version: %d\n", read_le16(buf));
+        return -1;
+    }
+
+    pmsg_info("serprog protocol version: %d\n", read_le16(buf));
+
+    // get programmer name
+    if (is_serprog_cmd_supported(my.cmd_bitmap, S_CMD_Q_PGMNAME)) {
+        memset(buf, 0, sizeof(buf));
+        if (perform_serprog_cmd(pgm, S_CMD_Q_PGMNAME, NULL, 0, buf, 16) != 0) {
+            pmsg_error("cannot get programmer name\n");
+            return -1;
+        }
+        pmsg_info("programmer name: %s\n", buf);
+    }
+
+    // check if required commands are supported
+    if (!is_serprog_cmd_supported(my.cmd_bitmap, S_CMD_O_SPIOP)) {
+        pmsg_error("the %s programmer does not support SPI operations\n", pgmid);
+        return -1;
+    }
+
+    if (!is_serprog_cmd_supported(my.cmd_bitmap, S_CMD_S_CS_MODE)) {
+        pmsg_error("the %s programmer does not support setting the CS mode\n", pgmid);
+        return -1;
+    }
+
+    if (!is_serprog_cmd_supported(my.cmd_bitmap, S_CMD_S_SPI_MODE)) {
+        pmsg_error("the %s programmer does not support setting the SPI mode\n", pgmid);
+        return -1;
+    }
+
+    if (my.cs > 0 && !is_serprog_cmd_supported(my.cmd_bitmap, S_CMD_S_SPI_CS)) {
+        pmsg_error("the %s programmer does not support changing the CS\n", pgmid);
+        return -1;
+    }
+
+    // set SPI clock frequency
+    if (is_serprog_cmd_supported(my.cmd_bitmap, S_CMD_S_SPI_FREQ)) {
+        memset(buf, 0, sizeof(buf));
+        write_le32(buf, 115200);
```

Comment:
Just got curious why the code chooses 115200 as SPI clock frequency. IIRC it's recommended to set the SPI clock frequency to no more than a quarter the CPU frequency of the to-be-programmed target chip. If the latter is factory reset it often is (but not always) 1 MHz (see `avrdude -p m328p/St | grep factory_fcpu`), so I'd expect sth like 250000 or 125000 but not 115200. Or perhaps better (if defined) `part->factory_fcpu/4`. `open()` doesn't have access to the part, so perhaps best to move setting the SPI clock frequency to `initialize()`?

Some AVR parts can be driven by vvv slow clocks, even as low as F<sub>CPU</sub> = 32768 Hz (eg, a clock quartz). In that case it would be critical that the programmer can lower its SPI clock. AVRDUDE lets the user specify the bitclock with `-B`, which is then made available to the `-c` programmer in `pgm->bitclock`. I think it would be better if that (or a value derived from that @MCUdude? I forgot whether bitclock is a frequency or time period) be used here instead of `part->factory_fcpu/4` if `pgm->bitclock` is non-zero?

Also, seeing that the programmer offers to set the SPI clock rate, `serprog.c` could offer to implement `pgm->set_sck_period()` and `pgm->get_sck_period()`. That would be *great*!

Reply:
I changed the implementation to use the specified bitclock and to fallback to `part->factory_fcpu/4` if available or 250000 if not. I also implemented `set_sck_period()` and `get_sck_period()`